src/verifier/Step5.sol

```diff
@@ -0,0 +1,650 @@
+    ) public view returns (KeccakTranscriptLib.KeccakTranscript memory, uint256) {
+        uint256[] memory input = compute_c_inner(proof);
+
+        uint8[] memory label = new uint8[](1);
+        label[0] = 0x65; // Rust's b"e"
+
+        transcript = KeccakTranscriptLib.absorb(transcript, label, input);
+
+        label[0] = 0x63; // Rust's b"c"
+
+        uint256 output;
+        (transcript, output) = KeccakTranscriptLib.squeeze(transcript, ScalarFromUniformLib.curvePallas(), label);
+
+        return (transcript, output);
+    }
+
+    function compute_powers_of_rho_primary(
+        Abstractions.RelaxedR1CSSNARK storage proof,
+        KeccakTranscriptLib.KeccakTranscript memory transcript
+    ) public view returns (KeccakTranscriptLib.KeccakTranscript memory, uint256[] memory) {
+        uint256 i = 0;
+        uint256[] memory evals = new uint256[](proof.eval_input_arr.length + proof.eval_output2_arr.length);
+
+        uint256 index;
+        for (index = 0; index < proof.eval_input_arr.length; index++) {
+            evals[i] = proof.eval_input_arr[index];
+            i++;
+        }
+        for (index = 0; index < proof.eval_output2_arr.length; index++) {
+            evals[i] = proof.eval_output2_arr[index];
+            i++;
+        }
+
+        uint8[] memory label = new uint8[](1);
+        label[0] = 0x65; // Rust's b"e"
+
+        transcript = KeccakTranscriptLib.absorb(transcript, label, evals);
+
+        label[0] = 0x72; // Rust's b"r"
+
+        // num_claims is hardcoded to be equal to 10
+        uint256 num_claims = 10;
+        uint256[] memory rho = new uint256[](num_claims);
+        (transcript, rho[0]) = KeccakTranscriptLib.squeeze(transcript, ScalarFromUniformLib.curveVesta(), label);
+
+        for (index = 1; index < num_claims; index++) {
+            rho[index] = mulmod(rho[index - 1], rho[0], Vesta.P_MOD);
+        }
+
+        return (transcript, rho);
+    }
+
+    function compute_powers_of_rho_secondary(
+        Abstractions.RelaxedR1CSSNARK storage proof,
+        KeccakTranscriptLib.KeccakTranscript memory transcript
+    ) public view returns (KeccakTranscriptLib.KeccakTranscript memory, uint256[] memory) {
+        uint256 i = 0;
+        uint256[] memory evals = new uint256[](proof.eval_input_arr.length + proof.eval_output2_arr.length);
+
+        uint256 index;
+        for (index = 0; index < proof.eval_input_arr.length; index++) {
+            evals[i] = proof.eval_input_arr[index];
+            i++;
+        }
+        for (index = 0; index < proof.eval_output2_arr.length; index++) {
+            evals[i] = proof.eval_output2_arr[index];
+            i++;
+        }
+
+        uint8[] memory label = new uint8[](1);
+        label[0] = 0x65; // Rust's b"e"
+
+        transcript = KeccakTranscriptLib.absorb(transcript, label, evals);
+
+        label[0] = 0x72; // Rust's b"r"
+
+        // num_claims is hardcoded to be equal to 10
```

Comment:
is there a way to relax this requirement?

Reply:
Not sure if I follow. This is hardcoded in the [reference](https://github.com/lurk-lab/Nova/blob/solidity-verifier-pp-spartan/src/spartan/ppsnark.rs#L1651) implementation as well.

Reply:
Oh, yes, I'm a dunce, and yes, the link to the reference Rust would definitely have helped silly me figure it out!